# Bioinformatics III: qiime2
---UPDATED June 20200 


###### tags: `MAE`, `qiime2`


overview of Qiime2 pipeline
https://docs.qiime2.org/2020.8/tutorials/overview/ 

![](https://i.imgur.com/s3VM4Gv.png)


You have an overview of amplicon sequencing here (Yates, 2016)
This book chapter covers DNA extraction, PCR, sequencing technologies and data analysis: https://drive.google.com/file/d/1zKUlOYZefoP3ZAGHv32cb7iROuHm3tgd/view?usp=sharing



[TOC]


:::danger


CONECTING to MBL remotely

From off campus it would be

ssh user@class.mbl.edu

and from there just call 
ssh class-02
:::

## Getting your terminal ready

## Getting your terminal ready

Navigate to the folder where you will run your analysis. 
Hint, you will need to use `cd` `ls` `pwd`

**here is the Unix tutorial of this course.** 
https://jbpc.mbl.edu/unix-tutorial/MAE2020.html

if you have *any doubt* about UNIX most probably is already solved in the tutorial, but we are happy to help

Now, wait and think. 

**are you running in a screened session?**
Yes? great. 
No? here is a brief how to. 

**summary for screen**
Screen
---

check here: [screen instructions](https://hackmd.io/Y3AabwWoTpObumF4eYb37g?view#screen)

navigate to the folder where you will run your analysis

1. to give a name to a screen session `screen -S sessionID`
2. List screens? `screen -r`
3. to connect to screen session `screen -r sessionID`
4. generate a file to log in specific parts of the session
Once in a screened session type `crt+a` release and press `:` a line of text will appear on the botton of the terminal Type `logfile NAMEforlog.txt` press enter (twice).  If you want to log something, press `crt+a` followed by `h`  in the botton of the terminal a line saying creating file NAMEforlog.txt will appear, press enter all you type now will be recorded. To stop recording type again `crt+a` followed by `h`. Recording will stop. **Supercool!** if you use `crt+a` followed by `h` it will append (instead of overwrite!) to the log file. To stop, once again  `crt+a` followed by `h`
4. close screen `crt+a` release and press `d`
5. kill screen `crt+a` release and press `k`
6. finished? type `exit`



## QIIME2 with SRA data

navigate to 
```
cd ~/saltmarsh/

```

Use the files downloaded [here](https://hackmd.io/@elperedoStudents/SkrHIBzSw)

After the loop, you should have downloaded 26 samples in .fastq.gz format. 
For each of them, you should have the R1 and R2 files, containing the paired-end reads. 

Also, you should have checked the quality of the reads summarized in the file .html (transfer to your computer using filezilla and open in any web browser)

The first graph plots the average phed score per position. The higher the number the better, as the probability of misscalling the base decreases. 
Also, you can "see" in the X axis how long the reads are. 

Do the reads we downloaded look like this?

![](https://i.imgur.com/efwejoL.png)


or like this?

![](https://i.imgur.com/5wLtBH2.png)





## Qiime2 tutorial
Here you can find the original tutorial: https://docs.qiime2.org/2020.2/tutorials/atacama-soils/ 

### What is Qiime2?

QIIME 2 is a powerful, extensible, and decentralized microbiome analysis package with a focus on data and analysis transparency. QIIME 2 enables researchers to start an analysis with raw DNA sequence data and finish with publication-quality figures and statistical results.

Key features:

Integrated and automatic tracking of data provenance

Semantic type system

Plugin system for extending microbiome analysis functionality

Support for multiple types of user interfaces (e.g. API, command line, graphical)

https://docs.qiime2.org/2020.2/concepts/

### Key vocabulary. 

##### Data files: QIIME 2 artifacts
Data produced by QIIME 2 exist as QIIME 2 artifacts. 

A QIIME 2 artifact contains **data and metadata**. The metadata describes things about the data, such as its type, format, and how it was generated (provenance). A QIIME 2 artifact typically has the `.qza` file extension when stored in a file.

Since QIIME 2 works with artifacts instead of data files the first thing to do is to import your data. Typically you will start by importing raw sequence data. 


#### Data files: visualizations
Visualizations are another type of data generated by QIIME 2. When written to disk, visualization files typically have the `.qzv` file extension. 


Use https://view.qiime2.org to easily view QIIME 2 visualizations files. 
Open the website, transfer the file to your computer, and drag and drop!


## Run Qiime2

First let's create a folder to run your data. To keep thind organized I would recommend to name it as follows:

qiime2_(use the project name)_and your initials!

```
mkdir qiime2_PRJNA610907
cd qiime2_PRJNA610907
 
```
the pwd would be ~/saltmarsh/qiime2_PRJNA610907

### Qiime2 files
Now, we need to prepare two files for running the program one lists the samples and tells the program "where" they are (file manifest). 
Int his file, the first column is the sample ID, the next are the path to the files. 

The second file (sample metadata) provides the information relevant for the analysis. Here you will list things like the treatments, time in culture, type of body part, or any variable that can be relevant to interpretate the data. In the first column we will list the sample ids (this has to match the ID in the manifest file). Then, each column is a different variable. 
The first line names the variables, the second line of the file lets the program if we are listing into a categorical or numerical variable. 

[metadata](https://earthmicrobiome.org/protocols-and-standards/metadata-guide/)  

#### Where are your samples?: “Fastq manifest” formats

Import your data into QIIME 2 manually by first creating a “manifest file” 

1.  You’ll create a text file called a manifest file named `pe-33-manifest.txt`, which maps sample identifiers to fastq files (or fastq.gz). 

2. The manifest file also indicates the direction of the reads (R1, R2).

3. The manifest file is a tab-seperated (i.e., .tsv) text file. The first column defines the Sample ID, while the second (and optional third) column defines the absolute filepath to the forward (and optional reverse) reads. All of the rules and behavior of this format are inherited from the QIIME 2 Metadata format.


Note: I usually create these files in excel and **then edit the format in a txt file**. 

Use the file SRR_Acc_List_PRJNA610907.txt which has all the sample names. 

[example of how to list samples using the SRR_list file](https://docs.google.com/spreadsheets/d/10QpoBROwj-kB9DvfaSkhfMT6IxC5DfL6MihcohfJi64/edit?usp=sharing)

Example of [final file ](https://docs.google.com/spreadsheets/d/1vCiXxT0CfZyQ2UQxDAQqAkFYlBbg9jmgYpR7_W0l0I4/edit?usp=sharing)

then just edit accordingly, you will need to add the full path to your samples is for example `/users/USERNAME/saltmarsh/SRAdump_PRJNA610907/`

So my file will look like

:::info
sample-id	forward-absolute-filepath	reverse-absolute-filepath
SRR11277344	/users/eperedo/saltmarsh/SRAdump_PRJNA610907/SRR11277348_pass_paired_1.fastq.gz	/users/eperedo/saltmarsh/SRAdump_PRJNA610907/SRR11277348_pass_paired_2.fastq.gz
:::

Once you have all the info in a txt file in your computer, we are ready to create the file in the server. 


1. You can transfer the file using filezilla. 
2. you can create the file using command line. 
To create the file, type in the terminal to create the txt. file. 

```
cat > pe-33-manifest.txt
``` 
And now let's populate it with the info you have prepared. 

* Copy from the file on your computer and paste in the terminal. Press enter to add an extra line.
* Close the file using Ctrl + Z

Let's check all looks correct!
```
ls -l pe-33-manifest.txt
head pe-33-manifest.txt
tail pe-33-manifest.txt
```

note. head and tail will by default give you 10 lines. If you want to see more use the flag number of lines `-n ` for example `-n 100` will list 100 lines. 

If something went wrong you can remove the file and start over!
```
rm pe-33-manifest.txt

```




### sample metadata 

Remember we downloaded the metadata from runselector clicking in Runinfo. Open the file in excel and select the columns you are interested. Then just format it as Qiime2 and save it in a txt file in your computer (name it sample-metadata.tsv). 

example [here](https://docs.google.com/spreadsheets/d/1wYQmje8_aDfWM05tkIHtDMpeLJzJbHQwY3DCPGZ_Byg/edit?usp=sharing)

As we did with the manifest file, once you have all the info in a txt file in your computer, we are ready to create the file in the server. 


1. You can transfer the file using filezilla. 
2. you can create the file using command line. 

As we did before, create a file 
```
cat > sample-metadata.tsv
```
And now let's populate it with the info you have prepared. 

* Enter your document's text. Copy from the file on your computer and past in the terminal. Press enter to add an extra line.
* Close the document using Ctrl + Z
and check!
```
ls -l sample-metadata.tsv
head sample-metadata.tsv
tail sample-metadata.tsv
```
if something went wrong you can remove the file and start over!
```
rm sample-metadata.tsv
```


## Running Qiime2

Let's activate the program!

Note the first time we have to configurate the profile and it is going to take a moment to load. In sucessive times we could use a more direct way of calling the progrma but for simplicity of instructions, I am listing the complete way of loading the program. 
note 2 (For Aster :P). All this derives from the conda installation that conflicts with some of the modules we load by default. 
```
module unload bioware
module load qiime2/2020.2.0-conda
conda init bash
. ~/.bashrc
conda activate /bioware/qiime2-2020.2.0-conda


```

to exit qiime2

```
conda deactivate

```


## Importing data into Qiime2

![](https://i.imgur.com/cjleTz9.png)

Using the manifest file we import the files of interest into QIIME2. These sequences are already demultiplexed, so it will import as "demux.qza"


:::info

**how you read the command**
*what* .... qiime tools import  
*how are my seqs?*   .... --type 'SampleData[PairedEndSequencesWithQuality]'   
*where are the seqs*  ...  --input-path pe-33-manifest.txt  
*How they look*--input-format PairedEndFastqManifestPhred33V2 
*outputs* ...--output-path paired-end-demux.qza  

:::
:::success
qiime tools import \
    --type 'SampleData[PairedEndSequencesWithQuality]' \
    --input-path pe-33-manifest.txt  \
    --output-path paired-end-demux.qza \
    --input-format PairedEndFastqManifestPhred33V2 
:::

Run in terminal

```
qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' --input-path pe-33-manifest.txt --output-path paired-end-demux.qza  --input-format PairedEndFastqManifestPhred33V2

```
now we are going to generate a human readable summary (we will do this after many steps, first we compute the data, then we summarinze)
:::success
qiime demux summarize \
    --i-data paired-end-demux.qza  \
    --o-visualization paired-end-demux.qzv
:::

Run in the terminal
```
qiime demux summarize --i-data paired-end-demux.qza --o-visualization paired-end-demux.qzv
```

using filezilla, drop the file qzv into your local computer. 
you can visualize the results in https://view.qiime2.org/ 

  
## Filtering and Clustering

![](https://i.imgur.com/85h8SCt.png)



**Remember that David covered the clustering of reads in this lecture.** https://drive.google.com/file/d/12qB3DK5M1GrTYFQU2w41iE1PQyedwvLU/view?usp=sharing


Once you have analyzed your sequence quality, you are ready to cluster yor sequences. 

We are going to use DADA2 to 
* remove low quality reads and pairs
* merge
* cluster
* remove quimeras


**What is sequence clustering?**

Grouping similar DNA sequences together – those that are of the same microbial “species”, called amplicon sequence variants (ASVs) or operational taxonomic units (OTUs) in the microbial world

For each group of similar DNA sequences, we are going to get representative sequences so that each amplicon sequence variant (ASV) is represented by a single DNA sequence (file = rep-seqs.qza)

We are also going to use this step to remove bad quality data. The program is going to compute the error frequenciest to remove those sequences that are different **as consequence of** sequencing errors (do you remember what David explained about singletons?). Finally we will also remove “junk” sequences (such as chimeric sequences and low quality). Chimeric sequence = combination of multiple erroneously joined DNA sequences (can be an artifact of PCR amplification). Singleton = a single, unique DNA sequence… most likely a sequencing error

After sequence clustering, you will have a table of samples by microbial species, with abundances (# of each type of bacterial sequence in each sample). This is called "table.qza"


Now, you are ready to run DADA2. Note that this step takes the longest (so again, are you running in screen) because it is very computationally intensive... it can take many hours to run, depending on how much data you have!

more info: here there is a link to the R package of dada2: https://benjjneb.github.io/dada2/tutorial.html



### How to modify the parameters in Dada2. 

We are using the file were we set all reads. 'paired-end-demux.qza'
The region v4-v5 is around 390 nt in length. We have reads that are 250 nt. We are going to trim to 220 (usually the tails of the reads have the lowest quality) so the overlap between the reads when we merge is still 30 nt 

for more info see here: https://docs.qiime2.org/2020.6/plugins/available/dada2/denoise-paired/ 
(other parameter of interest: --p-trim-left-f #number   --p-trim-left-r #number)

:::success
qiime dada2 denoise-paired  \
    --i-demultiplexed-seqs paired-end-demux.qza   \
    --p-trunc-len-f 220  \
    --p-trunc-len-r 220  \
    --o-table table.qza \
    --o-representative-sequences rep-seqs.qza   \
    --o-denoising-stats denoising-stats.qza   \
    --verbose  --p-n-threads 4
:::

Run in the terminal

```
qiime dada2 denoise-paired  --i-demultiplexed-seqs paired-end-demux.qza --p-trunc-len-f 220   --p-trunc-len-r 220  --o-table table.qza --o-representative-sequences rep-seqs.qza  --o-denoising-stats denoising-stats.qza  --verbose  --p-n-threads 4

```


##  Summarize the resutls

Next, create summary tables of clustered sequences.

:::success
qiime feature-table summarize  \
    --i-table table.qza \
    --o-visualization table.qzv \
    --m-sample-metadata-file sample-metadata.tsv
:::
Run in the terminal

```
qiime feature-table summarize  --i-table table.qza --o-visualization table.qzv --m-sample-metadata-file sample-metadata.tsv
```
:::success
qiime feature-table tabulate-seqs \
    --i-data rep-seqs.qza  \
    --o-visualization rep-seqs.qzv
:::
Run in the terminal
```
qiime feature-table tabulate-seqs --i-data rep-seqs.qza  --o-visualization rep-seqs.qzv
```
**stats, how many reads were removed because of low quality** This directly depends on the parameters you set to run the program so if too many reads are removed, you might want to double check those parameters. 

:::success
qiime metadata tabulate   \
    --m-input-file denoising-stats.qza  \
    --o-visualization denoising-stats.qzv
:::
Run in the terminal
```
qiime metadata tabulate   --m-input-file denoising-stats.qza  --o-visualization denoising-stats.qzv

```
Look at the “Overview” tab. How many samples do you have? How many unique sequences are there? How many total sequences? 

Go to the “Interactive Sample Detail” tab and view the number of sequences (“features”) in each sample. Use the sliding bar to see how many samples you will retain at each sampling depth. Do you need to exclude any samples?

Look carefully to the denoising stats. If too many seqs are lost and the quality of the reads do not indicate major issues you might need to adjust the dada2 parameters. Are you truncating the sequencies too much? too little?


## Taxonomic assignment

![](https://i.imgur.com/4GxuhNm.png)


We need to download the refence taxonomy databases that we will be using to identify the taxonomy of our ASV
silva, https://www.arb-silva.de/
green genes, https://greengenes.secondgenome.com/

We are going to use SIlVA. We are going to download a file that contains the info for the region of the 16S we are using. 

First we are going to create a folder to store the database nad navigate into it. 

```
mkdir ~/taxonomy_databases/
cd ~/taxonomy_databases/

```

now we will download the file from the command line. 


region V4V5
```
wget  -O "silva-132-99-515-806-nb-classifier.qza" "https://data.qiime2.org/2020.2/common/silva-132-99-515-806-nb-classifier.qza"
```
Full 16S sequence


```
wget  -O "silva-132-99-nb-classifier.qza" "https://data.qiime2.org/2020.2/common/silva-132-99-nb-classifier.qza"
```

**very important** your database has to match the version of your qiime2 installation (in this case 20202.2)

[links to other databases-- full lenght](https://docs.qiime2.org/2020.2/data-resources/#taxonomy-classifiers-for-use-with-q2-feature-classifier), e.g. green genes.
wget  -O "gg-13-8-99-nb-classifier.qza" "https://data.qiime2.org/2020.2/common/gg-13-8-99-nb-classifier.qza"

https://docs.qiime2.org/2020.2/data-resources/#taxonomy-classifiers-for-use-with-q2-feature-classifier


Now, we are ready to classify the sequences with the **full** silva database:

:::success
qiime feature-classifier classify-sklearn  \
    --i-classifier ~/taxonomy_databases/silva-132-99-nb-classifier.qza   \
    --i-reads rep-seqs.qza   \
    --o-classification taxonomy.qza 
    --p-n-jobs 6
:::

In the terminal
```
qiime feature-classifier classify-sklearn  --i-classifier ~/taxonomy_databases/silva-132-99-nb-classifier.qza  --i-reads rep-seqs.qza  --o-classification taxonomy.qza --p-n-jobs 6
```


**database for V4V5 region**
:::success
qiime feature-classifier classify-sklearn  \
    --i-classifier ~/taxonomy_databases/silva-132-99-515-806-nb-classifier.qza   \
    --i-reads rep-seqs.qza   \
    --o-classification taxonomy.qza --p-n-jobs 6
:::

In the terminal
```
qiime feature-classifier classify-sklearn  --i-classifier ~/taxonomy_databases/silva-132-99-515-806-nb-classifier.qza  --i-reads rep-seqs.qza  --o-classification taxonomy.qza 
```

and summarize

:::success
qiime metadata tabulate  \
    --m-input-file taxonomy.qza  \
    --o-visualization taxonomy.qzv 
:::

In the terminal
```
qiime metadata tabulate  --m-input-file taxonomy.qza --o-visualization taxonomy.qzv
```

Scan the taxonomy classifications. Which bacteria do you see? Any surprising or expected taxa? You can also use the search bar to find specific taxa.

lets do a graph to visualize the data

:::success
qiime taxa barplot 
    --i-table table.qza 
    --i-taxonomy taxonomy.qza 
    --m-metadata-file sample-metadata.tsv 
    --o-visualization taxa-bar-plots.qzv

:::

```
qiime taxa barplot  --i-table table.qza --i-taxonomy taxonomy.qza --m-metadata-file sample-metadata.tsv  --o-visualization taxa-bar-plots.qzv
```

Change the level (kingdom, phylum, class, order, family, genus, species) and order by different metadata variables… what trends do you see in the data?

You should see something like this:
![](https://i.imgur.com/EfcmAi1.png)

**Export ASV table**
:::success
qiime tools export   \
    --input-path table.qza  \
    --output-path ASV_table
:::

```
qiime tools export  --input-path table.qza   --output-path ASV_table

```

Convert the exported table from biom format to tsv

:::success
biom convert \
    -i ASV_table/feature-table.biom \
    -o ASV_table/feature-table.tsv \
    --to-tsv
:::

```
biom convert -i ASV_table/feature-table.biom -o ASV_table/feature-table.tsv --to-tsv

```
These tables can be used to generate Venn diagrams. 
Transfer the files to your computer, open the table in excel. 

Take all of the ASV names of the samples that you want to analyze, and put them into this Venn Diagram tool online:

https://bioinfogp.cnb.csic.es/tools/venny/

## Removal of mitochondrial and chloroplast 16S

Have you seen any eukaryotic sequences in the data? 
Why are they there? 

(hint 16S!)

Let's remove them by fitlering them out. 

:::success
qiime taxa filter-table   \
    --i-table table.qza   \
    --i-taxonomy taxonomy.qza   \
    --p-exclude mitochondria,chloroplast   \
    --o-filtered-table table-no-mitochondria-no-chloroplast.qza
:::
```
qiime taxa filter-table --i-table table.qza  --i-taxonomy taxonomy.qza --p-exclude mitochondria,chloroplast  --o-filtered-table table-no-mitochondria-no-chloroplast.qza
```


:::success
qiime feature-table summarize   \
    --i-table table-no-mitochondria-no-chloroplast.qza   \
    --o-visualization table-no-mitochondria-no-chloroplast.qzv   \
    --m-sample-metadata-file sample-metadata.tsv
:::
```
qiime feature-table summarize  --i-table table-no-mitochondria-no-chloroplast.qza  --o-visualization table-no-mitochondria-no-chloroplast.qzv  --m-sample-metadata-file sample-metadata.tsv
  
```

Now let's visualize the new filtered table
We will use the same scripts we used after dada2, just modifing the inputa table, create summary tables of clustered sequences.

Type:


:::success
qiime feature-table summarize   \
    --i-table table-no-mitochondria-no-chloroplast.qza   \
    --o-visualization table-no-mitochondria-no-chloroplast.qzv   \
    --m-sample-metadata-file sample-metadata.tsv
:::

```
qiime feature-table summarize  --i-table table-no-mitochondria-no-chloroplast.qza  --o-visualization table-no-mitochondria-no-chloroplast.qzv   --m-sample-metadata-file sample-metadata.tsv
```

:::success
qiime feature-table tabulate-seqs   \
    --i-data rep-seqs.qza  \
    --o-visualization rep-seqs-no-mitochondria-no-chloroplast.qzv
:::
```
qiime feature-table tabulate-seqs --i-data rep-seqs.qza --o-visualization rep-seqs-no-mitochondria-no-chloroplast.qzv
```

taxbar plots.
```
qiime taxa barplot  --i-table table-no-mitochondria-no-chloroplast.qza --i-taxonomy taxonomy.qza --m-metadata-file sample-metadata.tsv  --o-visualization taxa-bar-plots-no-mitochondria-no-chloroplast.qzv
```
**Export ASV table**
:::success
qiime tools export   \
    --input-path table.qza  \
    --output-path ASV_table
:::

```
qiime tools export  --input-path table-no-mitochondria-no-chloroplast.qza   --output-path ASV_table-no-mitochondria-no-chloroplast

```

Convert the exported table from biom format to tsv

:::success
biom convert \
    -i ASV_table/feature-table.biom \
    -o ASV_table/feature-table.tsv \
    --to-tsv
:::

```
biom convert -i ASV_table-no-mitochondria-no-chloroplast/feature-table.biom -o ASV_table-no-mitochondria-no-chloroplast/feature-table.tsv --to-tsv

```


### Metadata-based filtering


AKA HOW TO SELECT JUST PART of YOUR SAMPLES!

We might run DADA2 to cluster all types of samples from the  projects together... but now you just want to work on a subset? Get rid of those samples that you don't need to analyze! Otherwise they will appear in all of your graphs.

Now, select the samples that you want to analyze! For example, urchin people, select --p-where "Column of interest IN ('subset')" 

**SUPER IMPORTANT: From this point onwards, if you are interested in analyzing just a subset of the data you need to replace "table.qza" with "NAMEofSUBSET_table.qza"!**

:::success
qiime feature-table filter-samples  \
    --i-table table-no-mitochondria-no-chloroplast.qza   \
    --m-metadata-file sample-metadata.tsv   \
    --p-where "[ColumnName]='ANYTHING_YOU_WANT'"  \
    --o-filtered-table ANYTHING_YOU_WANT-table-no-mitochondria-no-chloroplast.qza
    
:::

```
qiime feature-table filter-samples  --i-table table-no-mitochondria-no-chloroplast.qza  --m-metadata-file sample-metadata.tsv   --p-where "[ColumnName]='ANYTHING_YOU_WANT'"  --o-filtered-table ANYTHING_YOU_WANT-table-no-mitochondria-no-chloroplast.qza

```

If there are multiple values that should be retained from a single metadata column, the IN clause can be used to specify those values. For example, the following command can be used to retain all samples with Ulva and Seawater.

*edited on Nov 29 to fix syntax in the --p-where flag* 
:::success
qiime feature-table filter-samples    \
      --i-table table-no-mitochondria-no-chloroplast.qza   \
      --m-metadata-file sample-metadata.tsv   \
      --p-where "[ColumnName] IN ('ANYTHING_YOU_WANT01','ANYTHING_YOU_WANT02')"   \
      --o-filtered-table ANYTHING_YOU_WANT_01_02--table-no-mitochondria-no-chloroplast.qza
:::

```
  qiime feature-table filter-samples   --i-table table-no-mitochondria-no-chloroplast.qza   --m-metadata-file sample-metadata.tsv   --p-where "[ColumnName] IN ('ANYTHING_YOU_WANT01','ANYTHING_YOU_WANT02')"  --o-filtered-table ANYTHING_YOU_WANT_01_02--table-no-mitochondria-no-chloroplast.qza
  
```


## rarefaction

**now we have a bacterial dataset, we will rarify**

[Rarefaction](https://en.wikipedia.org/wiki/Rarefaction_(ecology)) (ecology)

From Wikipedia, the free encyclopedia

In ecology, rarefaction is a technique to assess [species richness](https://en.wikipedia.org/wiki/Species_richness) [how many especies in a given sample] from the results of sampling. Rarefaction allows the calculation of species richness for a given number of individual samples, based on the construction of so-called rarefaction curves. This curve is a plot of the number of species as a function of the number of samples. Rarefaction curves generally grow rapidly at first, as the most common species are found, but the curves plateau as only the rarest species remain to be sampled.

The issue that occurs when sampling various species in a community is that the larger the number of individuals sampled, the more species that will be found. Rarefaction curves are created by randomly re-sampling the pool of N samples multiple times and then plotting the average number of species found in each sample (1,2, ... N). "Thus rarefaction generates the expected number of species in a small collection of n individuals (or n samples) drawn at random from the large pool of N samples.".


 **how we do that?**
 
 alpha- rarefaction. 
 We have defined the rarefaction. the alpha? just using the diversity (alpha diversity) of the sample. 
 
 From wiki.
 In ecology, alpha diversity ([α-diversity](https://en.wikipedia.org/wiki/Alpha_diversity)) is the mean species diversity in sites or habitats at a local scale. The term was introduced by R. H. Whittaker together with the terms beta diversity ([β-diversity](https://en.wikipedia.org/wiki/Beta_diversity)) and gamma diversity (γ-diversity). Whittaker's idea was that the total species diversity in a landscape ([gamma diversity](https://en.wikipedia.org/wiki/Gamma_diversity)) is determined by two different things, the mean species diversity in sites or habitats at a more local scale (alpha diversity) and the differentiation among those habitats (beta diversity).

[alpha-rarefaction](https://docs.qiime2.org/2020.2/plugins/available/diversity/alpha-rarefaction/?highlight=maximum%20rarefaction%20depth)
 
 First we group the sequences using a phylogenetic tree. Then we plot the curve of the number of detected species vs sampling depth
 
### Generate a phylogenetic tree. 
Type:
 A phylogenetic tree is a diagram that represents the evolutionary relationships among organisms:
![](https://i.imgur.com/Ifb0PJC.png)

Rooted trees contain a root at the base of the tree, which represents the common ancestor to all of the branches, while unrooted trees display relationships among taxa without assumptions about common ancestry"
![](https://i.imgur.com/qidd35j.png)

:::success
qiime phylogeny align-to-tree-mafft-fasttree \
--i-sequences rep-seqs.qza  \
--o-alignment aligned-rep-seqs.qza  \
--o-masked-alignment masked-aligned-rep-seqs.qza  
--o-tree unrooted-tree.qza  \
--o-rooted-tree rooted-tree.qza \
--verbose
:::

run in the terminal
```
qiime phylogeny align-to-tree-mafft-fasttree  --i-sequences rep-seqs.qza  --o-alignment aligned-rep-seqs.qza  --o-masked-alignment masked-aligned-rep-seqs.qza  --o-tree unrooted-tree.qza  --o-rooted-tree rooted-tree.qza --verbose --p-n-threads 6

```

### Generate a rarefaction curve 

We are going to visualize at what sequencing depth your taxa discovery reaches saturation

This command randomly subsamples your dataset at n = 1000, n = 2000, n = 3000... n = 50,000 sequences, and it plots the # of unique microbial taxa at each sequencing depth

:::success
qiime diversity alpha-rarefaction  \
--i-table table-no-mitochondria-no-chloroplast.qza \
--i-phylogeny rooted-tree.qza  \
--p-max-depth 35000  \
--m-metadata-file sample-metadata.tsv  \
--o-visualization alpha-rarefaction.qzv
:::

```
qiime diversity alpha-rarefaction   --i-table table-no-mitochondria-no-chloroplast.qza --i-phylogeny rooted-tree.qza  --p-max-depth 35000  --m-metadata-file sample-metadata.tsv  --o-visualization alpha-rarefaction.qzv

```
Your visualization should look something like this:

![](https://i.imgur.com/KT8kR6u.png)

Use this picture to decide the sequencing depth at which you want to “cut off” and standardize across all samples. (what number? the curve is flat and you retain as many samples as possible. the most important part, **the curve is flat**)
At which sequencing depth does the # of observed microbial species reach saturation?
Decide on a sequence cutoff that will maximize microbial sequence diversity and retain as many samples as possible!



[rarify in qiime2](https://docs.qiime2.org/2020.2/plugins/available/feature-table/rarefy/?highlight=rarefy)

We use this parameter in the command. 
--p-sampling-depth INTEGER
    Range(1, None)     The total frequency that each sample should be
                       rarefied to. Samples where the sum of frequencies is
                       less than the sampling depth will be not be included in
                       the resulting table unless subsampling is performed
                       with replacement.
                       
                       
Usually 10000 features is enough. We are going use this parameter across several datasets so we can compare later on different analysis. 

:::success
qiime feature-table rarefy \
    --i-table table-no-mitochondria-no-chloroplast.qza \
    --p-sampling-depth 10000 \
    --o-rarefied-table table-no-mitochondria-no-chloroplast-rarified.qza
:::

```
qiime feature-table rarefy  --i-table table-no-mitochondria-no-chloroplast.qza --p-sampling-depth 10000 --o-rarefied-table table-no-mitochondria-no-chloroplast-rarified.qza
```

:::success
qiime feature-table summarize  \
    --i-table table-no-mitochondria-no-chloroplast.qza  \
    --o-visualization table-no-mitochondria-no-chloroplast-rarified.qza  \
    --m-sample-metadata-file sample-metadata.tsv
:::

```
qiime feature-table summarize  --i-table table-no-mitochondria-no-chloroplast.qza  --o-visualization table-no-mitochondria-no-chloroplast-rarified.qza --m-sample-metadata-file sample-metadata.tsv

```


## Visualization of taxonomy after chloroplast removal and rarifaction.
 
 :::success
 qiime taxa barplot   \
 --i-table table-no-mitochondria-no-chloroplast-rarified.qza  \
 --i-taxonomy taxonomy.qza   \
 --m-metadata-file sample-metadata.tsv  \
 --o-visualization taxa-bar-plots-no-mitochondria-no-chloroplast-rarified.qzv
 
 :::
 
 ``` 
 qiime taxa barplot   --i-table table-no-mitochondria-no-chloroplast-rarified.qza   --i-taxonomy taxonomy.qza   --m-metadata-file sample-metadata.tsv   --o-visualization taxa-bar-plots-no-mitochondria-no-chloroplast-rarified.qzv
 
 ```
 
 Change the level (kingdom, phylum, class, order, family, genus, species) and order by different metadata variables… what trends do you see in the data?

You should see something like this:
![](https://i.imgur.com/EfcmAi1.png)


**Export ASV table** for doing Venn diagrams-- need to test this. 
```
qiime tools export   --input-path table-no-mitochondria-no-chloroplast-rarified.qza  --output-path ASV_table-no-mitochondria-no-chloroplast-rarified
```

Convert the exported table from biom format to tsv
```
biom convert -i ASV_table-no-mitochondria-no-chloroplast-rarified/feature-table.biom -o ASV_table-no-mitochondria-no-chloroplast-rarified/feature-table.tsv --to-tsv
```

Once you transfer the files to your computer, open the table in excel. 

Take all of the ASV names of the samples that you want to analyze, and put them into this Venn Diagram tool online:

https://bioinfogp.cnb.csic.es/tools/venny/

### **Export ASV table NOT RARIFIED ** for doing EDGE analysis. 
```
qiime tools export   --input-path table-no-mitochondria-no-chloroplast.qza  --output-path ASV_table-no-mitochondria-no-chloroplast
```

Convert the exported table from biom format to tsv
```
biom convert -i ASV_table-no-mitochondria-no-chloroplast/feature-table.biom -o ASV_table-no-mitochondria-no-chloroplast/feature-table-not-rarified.tsv --to-tsv
```

## Statistical Analysis

Alpha diversity (α) = the # of species at a local site

Beta diversity (β) = the turnover in species diversity among sites, calculated as:

β = (gamma – alpha) / alpha, where alpha = the average local diversity

Gamma diversity (γ) = the # of species within a region

## Diversity core metrics

Note: use the same --p sampling depth as you did before!!
:::success
qiime diversity core-metrics-phylogenetic   \
--i-phylogeny rooted-tree.qza   \
--i-table table-no-mitochondria-no-chloroplast-rarified.qza \
--p-sampling-depth 10000  \
--m-metadata-file sample-metadata.tsv  \
--output-dir core-metrics-results
:::
run in the terminal

```
qiime diversity core-metrics-phylogenetic   --i-phylogeny rooted-tree.qza   --i-table table-no-mitochondria-no-chloroplast-rarified.qza --p-sampling-depth 10000  --m-metadata-file sample-metadata.tsv  --output-dir core-metrics-results

```

[Jaccard index](https://en.wikipedia.org/wiki/Jaccard_index): also known as Intersection over Union and the Jaccard similarity coefficient (originally given the French name coefficient de communauté by Paul Jaccard), is a statistic used for gauging the similarity and diversity of sample sets. The Jaccard coefficient measures similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets:

[bray curtis index](https://en.wikipedia.org/wiki/Bray%E2%80%93Curtis_dissimilarity) In ecology and biology, the Bray–Curtis dissimilarity is a statistic used to quantify the compositional dissimilarity between two different sites, based on counts at each site. 
Unwieghted unifrac

[Uni frac](https://en.wikipedia.org/wiki/UniFrac) UniFrac is a distance metric used for comparing biological communities. It differs from dissimilarity measures such as Bray-Curtis dissimilarity in that it incorporates information on the relative relatedness of community members by incorporating phylogenetic distances between observed organisms in the computation. Both weighted (quantitative) and unweighted (qualitative) variants of UniFrac are widely used in microbial ecology, where the former accounts for abundance of observed organisms, while the latter only considers their presence or absence. The method was devised by Catherine Lozupone, when she was working under Rob Knight of the University of Colorado at Boulder in 2005.



### Alpha diversity

**Observed OTUs**: Calculates number of distinct OTUs

[OTU](https://en.wikipedia.org/wiki/Operational_taxonomic_unit)= the term Operational taxonomic unit "OTU" is also used in a different context and refers to clusters of (uncultivated or unknown) organisms, grouped by DNA sequence similarity of a specific taxonomic marker gene (originally coined as mOTU; molecular OTU). In other words, OTUs are pragmatic proxies for "species" (microbial or metazoan) at different taxonomic levels, in the absence of traditional systems of biological classification as are available for macroscopic organisms. For several years, OTUs have been the most commonly used units of diversity, especially when analysing small subunit 16S (for prokaryotes) or 18S rRNA (for eukaryotes) marker gene sequence datasets.

**Shannon’s diversity index**: [more here](https://en.wikipedia.org/wiki/Diversity_index#Shannon_index). measure of community richness, accounts for both abundance and evenness of the taxa present

**Faith’s Phylogenetic Diversity**: [more here](https://methodsblog.com/tag/faiths-phylogenetic-diversity/) measure of community richness that incorporates phylogenetic relationships between the features (sum of length of branches)

**Pielou’s Evenness**: [more here](https://en.wikipedia.org/wiki/Species_evenness) a measure of community evenness


Conduct alpha diversity statistics in QIIME2

### observed OTUS

```
qiime diversity alpha-group-significance   --i-alpha-diversity core-metrics-results/observed_otus_vector.qza   --m-metadata-file sample-metadata.tsv   --o-visualization core-metrics-results/observed-otus-group-significance.qzv

```
### faith phylogenetic distance
```
qiime diversity alpha-group-significance  --i-alpha-diversity core-metrics-results/faith_pd_vector.qza  --m-metadata-file sample-metadata.tsv   --o-visualization core-metrics-results/faith_pd_vector-significance.qzv

```
### shannon diversity
```
qiime diversity alpha-group-significance  --i-alpha-diversity core-metrics-results/shannon_vector.qza  --m-metadata-file sample-metadata.tsv  --o-visualization core-metrics-results/shannon_vector-significance.qzv

```

### evenness
```
qiime diversity alpha-group-significance  --i-alpha-diversity core-metrics-results/evenness_vector.qza  --m-metadata-file sample-metadata.tsv   --o-visualization core-metrics-results/evenness_vector-significance.qzv

```

Change the column to select different metadata variables and look at the results… does the mean # of observed OTUs differ significantly by site? By the presence of a given variable? Other metadata variables?

Examples of alpha diversity analyses in published papers

![](https://i.imgur.com/XUb6OKY.png)
![](https://i.imgur.com/FPpHKzP.png)
![](https://i.imgur.com/az95WVe.png)

more cool graphs: https://forum.qiime2.org/t/tutorial-integrating-qiime2-and-r-for-data-visualization-and-analysis-using-qiime2r/4121


## Beta diversity

ASV table: microbial abundances in each sample  --> turns into:
Bray-Curtis similarity matrix: similarity of each sample to all other samples

Two types of B diversity plots:
1) PCoA plot: principal components analysis
2) NMDS plot: non-metric multidimensional scaling

### Visualizations for

 
The graphs are in the folder core-metrics-results/

Bray Curtis PCoA plot: bray_curtis_emperor.qzv 

Jaccard PCoA plot:  jaccard_emperor.qzv

Unweighted unifrac PCoA plot: unweighted_unifrac_emperor.qzv

Weighted unifrac PCoA plot: weighted_unifrac_emperor.qzv

:::info
What is weighted unifrac?
Communities are considered more related if the taxa they contain are more closely related

UniFrac measures the phylogenetic distance between sets of taxa in a tree (sum of branch lengths shared between samples)

Weighted UniFrac also accounts for the relative abundance of lineages between communities
:::
![](https://i.imgur.com/GF6QWNH.png)

1. Each dot = a 2D representation of one sample and the entire microbial community!
2. Use the drop-down menu to analyze with different metadata variables 

Change the column to select different metadata variables and look at the results… what trends do you see in the data?


## Beta diversity statistics in QIIME2

Change the  --m-metadata-column to run statistical tests with different variables: Does microbial community structure differ significantly among different samples according to X (site, vegetation…) ? 

REMEMBER CHANGING ####column### TO THE PROPER VARIABLE!!!

```
qiime diversity beta-group-significance  --i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza  --m-metadata-file sample-metadata.tsv  --m-metadata-column ####column###  --o-visualization core-metrics-results/unweighted-unifrac-####column###-significance.qzv  --p-pairwise

```
REady for more analysis?

https://docs.qiime2.org/2018.11/tutorials/overview/#diversity 

https://docs.qiime2.org/2018.11/tutorials/qiime2-for-experienced-microbiome-researchers/

https://grunwaldlab.github.io/analysis_of_microbiome_community_data_in_r/07--diversity_stats.html

Examples of beta diversity analyses in published papers

![](https://i.imgur.com/1Lfn2hm.png)
![](https://i.imgur.com/8ybbkD8.png)
![](https://i.imgur.com/BtfjGD9.png)


That’s all for now… but there are so many additional analyses that you can do!

![](https://i.imgur.com/Z6Be0ec.png)

See: https://docs.qiime2.org/

